src/twisted/web/test/test_http2.py

```diff
@@ -946,8 +946,9 @@ def test_terminatedRequest(self):
         self.assertTrue(request._disconnected)
         self.assertTrue(request.channel is None)
 
-        # An attempt to write should at this point raise an exception.
-        self.assertRaises(AttributeError, request.write, b"third chunk")
+        # This should not raise an exception, the function will just return
+        # immediately, and do nothing
+        request.write(b"third chunk")
```

Comment:
Hmm.  So this is actually a tested behavior of a public interface being changed...  :thinking: 

Although, jeez, `AttributeError`?  That's a pretty sad behavior.  I expect whoever wrote this test just encoded whatever accidental implementation behavior they happened to get without thinking about it very hard.

Yet, regardless of that, it's still a tested behavior of a public interface.  Strictly speaking, this is a backwards incompatible change (what if someone wrote an application that depends on `AttributeError` here to notice a closed connection?).

Ignoring backward compatibility concerns, I'd probably go along with this new behavior as sensible, but I don't think we get to ignore those concerns.  Unless... considering people are complaining about this behavior after upgrading, it could be the exception behavior is relatively new and was introduced accidentally as part of some other work?  

The "safe" thing for me to say is that this fix needs to be inside the wsgi implementation instead of inside the `IRequest` implementation.  If we knew exactly when the exception behavior was introduced, maybe we could make a different decision.

And, with a little help from git bisect, I see that ticket:8191 is where the change was introduced and it was first released in 16.3.0.  That's annoyingly long ago.  Does it mean the new behavior has to be considered the intended behavior and compatibility needs to be maintained?  Or do we still just call this a regression fix, even though the regression has been in the wild for 3 years & 10 releases...

Reply:
Yes it definitely is a behavior change (or more-precisely, change back?) and I suppose there could be projects out there that modified their behavior and then will have to modify it again. 

I think some kind of behavior change is called for though, whether or not we throw another kind of (more catcheable than AttributeError) exception or go back to the old behavior (defer exception until close()). This PR would be for the latter option.

The project I'm working on could go either way - we'd be happy to just catch and ignore a specific-enough exception.